### PR TITLE
gadgets: Rework the makefile to add better parallelism.

### PR DESCRIPTION
**Pull request by eiffel-fl**
Hi.

This PR reworks the gadget Makefile to use parallelism instead of `for`:

```bash
$ make sign -j$(nproc)
...
$ echo $?
0
$ sudo -E ../ig image list | grep sched francis/makefile-gadgets-parallel % u=
INFO[0000] Experimental features enabled                
francisacrregistry.azurecr.io/gadget/ci/sched_cls_drop francis-makefile-gadgets-parallel sha256:881f5946584002991586a65245bd726dcd7ffe55d051fe34a2f8797d998ad47b 2024-06-11T13:40:59+01:00
$ cosign verify --key ../cosign.pub francisacrregistry.azurecr.io/gadget/ci/sched_cls_drop@sha256:881f5946584002991586a65245bd726dcd7ffe55d051fe34a2f8797d998ad47b

Verification for francisacrregistry.azurecr.io/gadget/ci/sched_cls_drop@sha256:881f5946584002991586a65245bd726dcd7ffe55d051fe34a2f8797d998ad47b --
The following checks were performed on each of these signatures:
  - The cosign claims were validated
  - Existence of the claims in the transparency log was verified offline
  - The signatures were verified against the specified public key

[{"critical":{"identity":{"docker-reference":"francisacrregistry.azurecr.io/gadget/ci/sched_cls_drop"},"image":{"docker-manifest-digest":"sha256:881f5946584002991586a65245bd726dcd7ffe55d051fe34a2f8797d998ad47b"},"type":"cosign container image signature"},"optional":{"Bundle":{"SignedEntryTimestamp":"MEQCIAoiihYjJU3l+7ofdXNysKuN9pZbO6+dnAangxvMy4ieAiAflhyUGigzTepUD3isdPOsgHbyHbDBN8sj41+sYObIow==","Payload":{"body":"eyJhcGlWZXJzaW9uIjoiMC4wLjEiLCJraW5kIjoiaGFzaGVkcmVrb3JkIiwic3BlYyI6eyJkYXRhIjp7Imhhc2giOnsiYWxnb3JpdGhtIjoic2hhMjU2IiwidmFsdWUiOiI1MGQwNzY4MzlkMDA5ODA5MDNkMDg1N2U3ZjJhNzBhMjIyYzIwYTE3ZGViNTJmYzVkYWI5NTA1ODc1ZjU1ZTUyIn19LCJzaWduYXR1cmUiOnsiY29udGVudCI6Ik1FVUNJSDMveUdycG9vNlJsZE1vcTZvb1RzWlFDem42WTRub0dTR2ZtTlFqbXN0aUFpRUF2NWJZRW5KQjFSSmNRbEpFZkk2U3JqbXNhYzRVLzRreXJ6VHEwZWtBcWdnPSIsInB1YmxpY0tleSI6eyJjb250ZW50IjoiTFMwdExTMUNSVWRKVGlCUVZVSk1TVU1nUzBWWkxTMHRMUzBLVFVacmQwVjNXVWhMYjFwSmVtb3dRMEZSV1VsTGIxcEplbW93UkVGUlkwUlJaMEZGTms0MWVFSlRaR3BITTBnM1EwMU5iVnBFYm1wa1ExTjVZVXRTZFFwaWFraDZVbVF3VG5jMVZsbFJhRVpTU1hoWlNtMDVPV0ZUZGpVMmFqUlJiME5KYlhKQmJtNWtVRmN6ZG1SNFNWSkRhV1JTVUN0TWRXNTNQVDBLTFMwdExTMUZUa1FnVUZWQ1RFbERJRXRGV1MwdExTMHRDZz09In19fX0=","integratedTime":1718109683,"logIndex":101690243,"logID":"c0d23d6ad406973f9559f3ba2d1ca01f84147d8ffc5b8445c224f98b9591801d"}}}}]
```

Best regards.

### CODE REVIEW

**Review by mauriciovasquezbernal — APPROVED** [review]
Good addition. Thanks for improving it!

File: gadgets/Makefile [hunk 1]
```diff
@@ -77,25 +77,39 @@ $(GADGETS_README):
 		ln -sf ../../gadgets/$@ $(ROOT_DIR)../docs/gadgets/$(shell dirname $@).mdx ; \
 	fi
 
+%-push: %
+	@echo "Pushing $*"
+	@sudo -E IG_EXPERIMENTAL=true $(IG) image push $(GADGET_REPOSITORY)/$*:$(GADGET_TAG)
```
Comment (alban) [inline gadgets/Makefile:82]: IG_EXPERIMENTAL=true is not needed for pushing anymore.

Ditto below.

**Review by alban — COMMENTED** [review]
I think your solution with `sign-existing` and `push-existing` is good.

I was searching for alternatives with [.WAIT](https://www.gnu.org/software/make/manual/make.html#index-g_t_002eNOTPARALLEL-238) such as the following, but it makes it more confusing, so your solution is better.
```Makefile
build-and-push-and-sign: build .WAIT push .WAIT sign
```

File: gadgets/Makefile [hunk 2]
```diff
@@ -78,24 +78,34 @@ $(GADGETS_README):
...
-		sudo -E $(IG) image push $(GADGET_REPOSITORY)/$$GADGET:$(GADGET_TAG) || exit 1 ; \
-	done
-
-sign: push
-	@echo "Signing all gadgets"
-	for GADGET in $(GADGETS); do \
-		digest=$$(sudo -E $(IG) image inspect $(GADGET_REPOSITORY)/$$GADGET:$(GADGET_TAG) -o json | jq -r .Digest) ; \
-		cosign sign --key env://COSIGN_PRIVATE_KEY --yes --recursive $(GADGET_REPOSITORY)/$$GADGET@$$digest || exit 1 ; \
-	done
+push: $(addsuffix -push,$(GADGETS))
```
Comment (burak-ok) [inline gadgets/Makefile:81]: Isn't there a rule missing to build `%-push`?

**Comment by mauriciovasquezbernal** [comment]
It seems it still creates some issues on the CI: #3555

**Pull request closed without merging**